File: internal/service/vectorize.go
package service

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// VecParams are the user-selectable vectorization settings.
type VecParams struct {
	Mode            string // spline | polygon | pixel
	ColorPrecision  int    // 2-8
	FilterSpeckle   int    // 0-20
	CornerThreshold int    // 1-180
}

func (p *VecParams) normalize() {
	switch p.Mode {
	case "polygon", "pixel", "spline":
	default:
		p.Mode = "spline"
	}
	if p.ColorPrecision < 2 || p.ColorPrecision > 8 {
		p.ColorPrecision = 6
	}
	if p.FilterSpeckle < 0 || p.FilterSpeckle > 20 {
		p.FilterSpeckle = 4
	}
	if p.CornerThreshold < 1 || p.CornerThreshold > 180 {
		p.CornerThreshold = 60
	}
}

// ToolAvailability contains the status of external conversion tools.
type ToolAvailability struct {
	Vtracer   bool
	Inkscape  bool
	Potrace   bool
	Libre     bool
	Soffice   string
	InkscapeBin string
	PotraceBin  string
}

// DetectTools probes for external binaries.
func DetectTools() ToolAvailability {
	a := ToolAvailability{}
	if _, err := exec.LookPath("vtracer"); err == nil {
		a.Vtracer = true
	}
	if _, err := exec.LookPath("inkscape"); err == nil {
		a.Inkscape = true
		a.InkscapeBin = "inkscape"
	}
	if _, err := exec.LookPath("potrace"); err == nil {
		a.Potrace = true
		a.PotraceBin = "potrace"
	}
	for _, name := range []string{"soffice", "libreoffice"} {
		if _, err := exec.LookPath(name); err == nil {
			a.Libre = true
			a.Soffice = name
			break
		}
	}
	return a
}

// vectorizeImage converts a raster image to SVG via the Python wrapper.
func vectorizeImage(src, outSVG string, p VecParams) error {
	p.normalize()
	params := map[string]interface{}{
		"colormode":        "color",
		"mode":             p.Mode,
		"filter_speckle":   p.FilterSpeckle,
		"color_precision":  p.ColorPrecision,
		"corner_threshold": p.CornerThreshold,
		"path_precision":   8,
	}
	b, _ := json.Marshal(params)
	out, err := RunCmd(PythonPath(), ScriptPath("vectorize.py"), "svg", src, outSVG, string(b))
	if err != nil {
		return fmt.Errorf("vtracer 失败: %s", strings.TrimSpace(out))
	}
	if _, err := os.Stat(outSVG); err != nil {
		return fmt.Errorf("vtracer 未能生成SVG: %s", strings.TrimSpace(out))
	}
	return nil
}

// Vectorize performs the image -> target format conversion.
// Returns the path of the converted file.
func Vectorize(tmpDir, src, output string, p VecParams) (string, error) {
	// Normalize input to PNG first for vtracer compatibility
	pngPath := filepath.Join(tmpDir, "input.png")
	if out, err := RunCmd(PythonPath(), ScriptPath("vectorize.py"), "topng", src, pngPath); err != nil {
		return "", fmt.Errorf("图片读取失败: %s", strings.TrimSpace(out))
	}
	output = SafeExt(output)
	if output == "" {
		output = "svg"
	}

	// Produce the base SVG
	svgPath := filepath.Join(tmpDir, "out.svg")
	if err := vectorizeImage(pngPath, svgPath, p); err != nil {
		return "", err
	}

	switch output {
	case "svg":
		dest := filepath.Join(tmpDir, "result.svg")
		return dest, CopyFile(svgPath, dest)

	case "pdf", "ai", "eps":
		// Use Inkscape for PDF/EPS; AI is exported as PDF-compatible
		inkscape := DetectTools().InkscapeBin
		if inkscape == "" {
			return "", fmt.Errorf("inkscape 未安装，无法输出 %s 格式，请改用 SVG", strings.ToUpper(output))
		}
		// AI format: export as PDF first, then rename (AI is PDF-compatible)
		if output == "ai" {
			pdfPath := filepath.Join(tmpDir, "result.pdf")
			args := []string{"--export-type=pdf", "--export-filename=" + pdfPath, svgPath}
			cmd := exec.Command(inkscape, args...)
			out, err := cmd.CombinedOutput()
			if err != nil {
				return "", fmt.Errorf("inkscape 转换失败: %s", strings.TrimSpace(string(out)))
			}
			if _, err := os.Stat(pdfPath); err != nil {
				return "", fmt.Errorf("inkscape 输出文件缺失")
			}
			dest := filepath.Join(tmpDir, "result.ai")
			if err := CopyFile(pdfPath, dest); err != nil {
				return "", fmt.Errorf("复制 AI 文件失败")
			}
			return dest, nil
		}
		dest := filepath.Join(tmpDir, "result."+output)
		ext := output
		args := []string{"--export-type=" + ext, "--export-filename=" + dest, svgPath}
		cmd := exec.Command(inkscape, args...)
		out, err := cmd.CombinedOutput()
		if err != nil {
			return "", fmt.Errorf("inkscape 转换失败: %s", strings.TrimSpace(string(out)))
		}
		if _, err := os.Stat(dest); err != nil {
			// Fallback: inkscape might have named it differently
			alt := filepath.Join(tmpDir, "result."+output)
			if _, err := os.Stat(alt); err != nil {
				return "", fmt.Errorf("inkscape 输出文件缺失")
			}
			dest = alt
		}
		return dest, nil

	case "dxf":
		// Use Potrace for DXF (monochrome).
		potrace := DetectTools().PotraceBin
		if potrace == "" {
			return "", fmt.Errorf("potrace 未安装，无法输出 DXF 格式")
		}
		pbmPath := filepath.Join(tmpDir, "input.pbm")
		if out, err := RunCmd(PythonPath(), ScriptPath("vectorize.py"), "topbm", pngPath, pbmPath); err != nil {
			return "", fmt.Errorf("PBM 转换失败: %s", strings.TrimSpace(out))
		}
		dest := filepath.Join(tmpDir, "result.dxf")
		cmd := exec.Command(potrace, "-b", "dxf", "-o", dest, pbmPath)
		out, err := cmd.CombinedOutput()
		if err != nil {
			return "", fmt.Errorf("potrace 转换失败: %s", strings.TrimSpace(string(out)))
		}
		if _, err := os.Stat(dest); err != nil {
			return "", fmt.Errorf("potrace 输出文件缺失")
		}
		return dest, nil

	case "sk":
		dest := filepath.Join(tmpDir, "result.sketch")
		if out, err := RunCmd(PythonPath(), ScriptPath("vectorize.py"), "sketch", svgPath, dest); err != nil {
			return "", fmt.Errorf("Sketch 包装失败: %s", strings.TrimSpace(out))
		}
		return dest, nil

	case "fig":
		dest := filepath.Join(tmpDir, "result.fig")
		if out, err := RunCmd(PythonPath(), ScriptPath("vectorize.py"), "fig", svgPath, dest); err != nil {
			return "", fmt.Errorf("Figma 包装失败: %s", strings.TrimSpace(out))
		}
		return dest, nil

	default:
		return "", fmt.Errorf("不支持的输出格式: %s", output)
	}
}

File: src/service/vectorize.rs
use crate::util::{run_cmd, python_path, script_path, safe_ext};
use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct VecParams {
    pub mode: String,
    pub color_precision: i32,
    pub filter_speckle: i32,
    pub corner_threshold: i32,
}

impl Default for VecParams {
    fn default() -> Self {
        Self {
            mode: "spline".to_string(),
            color_precision: 6,
            filter_speckle: 4,
            corner_threshold: 60,
        }
    }
}

impl VecParams {
    pub fn normalize(&mut self) {
        match self.mode.as_str() {
            "polygon" | "pixel" | "spline" => {}
            _ => self.mode = "spline".to_string(),
        }
        if self.color_precision < 2 || self.color_precision > 8 {
            self.color_precision = 6;
        }
        if self.filter_speckle < 0 || self.filter_speckle > 20 {
            self.filter_speckle = 4;
        }
        if self.corner_threshold < 1 || self.corner_threshold > 180 {
            self.corner_threshold = 60;
        }
    }
}

#[derive(Debug, Clone)]
pub struct ToolAvailability {
    pub vtracer: bool,
    pub inkscape: bool,
    pub potrace: bool,
    pub libre: bool,
    pub soffice: String,
    pub inkscape_bin: String,
    pub potrace_bin: String,
}

pub fn detect_tools() -> ToolAvailability {
    let vtracer = which::which("vtracer").is_ok();
    let inkscape_bin = which::which("inkscape").map(|p| p.to_string_lossy().to_string()).unwrap_or_default();
    let inkscape = !inkscape_bin.is_empty();
    let potrace_bin = which::which("potrace").map(|p| p.to_string_lossy().to_string()).unwrap_or_default();
    let potrace = !potrace_bin.is_empty();
    let (libre, soffice) = ["soffice", "libreoffice"]
        .into_iter()
        .find_map(|name| {
            which::which(name).ok().map(|p| (true, p.to_string_lossy().to_string()))
        })
        .unwrap_or((false, String::new()));
    ToolAvailability {
        vtracer,
        inkscape,
        potrace,
        libre,
        soffice,
        inkscape_bin,
        potrace_bin,
    }
}

pub fn vectorize_image(src: &str, out_svg: &str, params: &VecParams) -> Result<(), String> {
    let mut p = params.clone();
    p.normalize();
    let params_json = serde_json::json!({
        "colormode": "color",
        "mode": p.mode,
        "filter_speckle": p.filter_speckle,
        "color_precision": p.color_precision,
        "corner_threshold": p.corner_threshold,
        "path_precision": 8,
    });
    let result = run_cmd(python_path(), &[
        &script_path("vectorize.py"),
        "svg",
        src,
        out_svg,
        &params_json.to_string(),
    ]);
    if let Some(ref err) = result.error {
        return Err(format!("vtracer 失败: {}", err));
    }
    if !std::path::Path::new(out_svg).exists() {
        return Err(format!("vtracer 未能生成SVG: {}", result.stdout.trim()));
    }
    Ok(())
}

pub fn vectorize(
    tmp_dir: &str,
    src: &str,
    output: &str,
    params: VecParams,
) -> Result<String, String> {
    // topng first
    let png_path = PathBuf::from(tmp_dir).join("input.png");
    let result = run_cmd(python_path(), &[
        &script_path("vectorize.py"),
        "topng",
        src,
        png_path.to_str().unwrap(),
    ]);
    if let Some(ref err) = result.error {
        return Err(format!("图片读取失败: {}", err));
    }

    let output = safe_ext(output);
    let output = if output.is_empty() { "svg" } else { &output };

    let svg_path = PathBuf::from(tmp_dir).join("out.svg");
    vectorize_image(png_path.to_str().unwrap(), svg_path.to_str().unwrap(), &params)?;

    match output {
        "svg" => {
            let dest = PathBuf::from(tmp_dir).join("result.svg");
            std::fs::copy(&svg_path, &dest).map_err(|e| format!("复制失败: {}", e))?;
            Ok(dest.to_string_lossy().to_string())
        }
        "pdf" | "ai" | "eps" => {
            let tools = detect_tools();
            if tools.inkscape_bin.is_empty() {
                return Err(format!("inkscape 未安装，无法输出 {} 格式，请改用 SVG", output.to_uppercase()));
            }
            let dest = PathBuf::from(tmp_dir).join(format!("result.{}", output));
            let export_type = if output == "ai" { "pdf" } else { output };
            let result = run_cmd(&tools.inkscape_bin, &[
                "--export-type", export_type,
                "--export-filename", dest.to_str().unwrap(),
                svg_path.to_str().unwrap(),
            ]);
            if let Some(ref err) = result.error {
                return Err(format!("inkscape 转换失败: {}", err));
            }
            if !dest.exists() {
                return Err("inkscape 输出文件缺失".to_string());
            }
            if output == "ai" {
                // For AI, copy the PDF as AI
                let ai_dest = PathBuf::from(tmp_dir).join("result.ai");
                std::fs::copy(&dest, &ai_dest).map_err(|e| format!("复制 AI 文件失败: {}", e))?;
                Ok(ai_dest.to_string_lossy().to_string())
            } else {
                Ok(dest.to_string_lossy().to_string())
            }
        }
        "dxf" => {
            let tools = detect_tools();
            if tools.potrace_bin.is_empty() {
                return Err("potrace 未安装，无法输出 DXF 格式".to_string());
            }
            let pbm_path = PathBuf::from(tmp_dir).join("input.pbm");
            let result = run_cmd(python_path(), &[
                &script_path("vectorize.py"),
                "topbm",
                png_path.to_str().unwrap(),
                pbm_path.to_str().unwrap(),
            ]);
            if let Some(ref err) = result.error {
                return Err(format!("PBM 转换失败: {}", err));
            }
            let dest = PathBuf::from(tmp_dir).join("result.dxf");
            let result = run_cmd(&tools.potrace_bin, &[
                "-b", "dxf",
                "-o", dest.to_str().unwrap(),
                pbm_path.to_str().unwrap(),
            ]);
            if let Some(ref err) = result.error {
                return Err(format!("potrace 转换失败: {}", err));
            }
            if !dest.exists() {
                return Err("potrace 输出文件缺失".to_string());
            }
            Ok(dest.to_string_lossy().to_string())
        }
        "sk" => {
            let dest = PathBuf::from(tmp_dir).join("result.sketch");
            let result = run_cmd(python_path(), &[
                &script_path("vectorize.py"),
                "sketch",
                svg_path.to_str().unwrap(),
                dest.to_str().unwrap(),
            ]);
            if let Some(ref err) = result.error {
                return Err(format!("Sketch 包装失败: {}", err));
            }
            Ok(dest.to_string_lossy().to_string())
        }
        "fig" => {
            let dest = PathBuf::from(tmp_dir).join("result.fig");
            let result = run_cmd(python_path(), &[
                &script_path("vectorize.py"),
                "fig",
                svg_path.to_str().unwrap(),
                dest.to_str().unwrap(),
            ]);
            if let Some(ref err) = result.error {
                return Err(format!("Figma 包装失败: {}", err));
            }
            Ok(dest.to_string_lossy().to_string())
        }
        _ => Err(format!("不支持的输出格式: {}", output)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_vec_params_normalize() {
        let mut p = VecParams {
            mode: "invalid".to_string(),
            color_precision: 1,
            filter_speckle: -1,
            corner_threshold: 0,
        };
        p.normalize();
        assert_eq!(p.mode, "spline");
        assert_eq!(p.color_precision, 6);
        assert_eq!(p.filter_speckle, 4);
        assert_eq!(p.corner_threshold, 60);
    }

    #[test]
    fn test_detect_tools() {
        let tools = detect_tools();
        // Just ensure it runs without panicking
        let _ = tools;
    }
}
